### PR TITLE
Sparse indices

### PR DESCRIPTION
**Pull request by yb6599**

### CODE REVIEW

File: pysindy/optimizers/stlsq.py [hunk 1]
```diff
@@ -125,28 +126,55 @@ def __init__(
...
+                    self.alpha = 2 * self.alpha
+            self.iters += 1
+            return coef
+        if sparse_ind is not None:
+            coef = np.zeros(dim)
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=LinAlgWarning)
+                try:
+                    coef_sparse = ridge_regression(x, y, self.alpha, **kw)
+                    coef_notsparse = ridge_regression(x, y, alpha=0, **kw)
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:167]: Don't use two separate regressions.  By way of explanation, consider the ridge regression objective for the case `sparse_ind=[1]`:

$$
\left\\|\left[\begin{matrix}1 & 1 \\\\ 0 & 1\end{matrix}\right] \left[\begin{matrix}x \\\\ y \end{matrix}\right] - \left[\begin{matrix}1 \\\\ 1\end{matrix}\right]\right\\|^2 + \left\\|\left[\begin{matrix}0 \\\\ y\end{matrix}\right]\right\\|^2.
$$

This is equivalent to the expression 

$$
(x+y-1)^2 + (y-1)^2 + y^2
$$

Your code instead finds $y$ from $(x+y-1)^2 + (y-1)^2 +x^2 + y^2$ and $x$ from $(x+y-1)^2 + (y-1)^2$ and concats them together, which isn't the same thing.  In the correct case, $x=y=.5$.  In the zero-penalty case, $x=0, y=1$.  In the full-penalty case, $0\lt x\lt.5,.5\lt y\lt1$.  your code would return $x=0, .5\lt y \lt 1$

The test only passes because our test has a diagonal matrix.  Instead, pass an array as the alpha parameter (See [the Ridge docs](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html)) and add the above equation as a new test.

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Good first pass!  However, you should set alpha to an array of weights in `ridge_regression`, rather than running two regressions.

File: pysindy/optimizers/stlsq.py (hunk 1 above)
Comment (yb6599) [reply pysindy/optimizers/stlsq.py:167]: I understand why it does not work for non diagonal matrices.

However, when I try to pass an array as the alpha parameter, it does not work because:
While we are trying to minimize $\||Ax-b\||^2_2 + \alpha \||x\||^2_2$, the shape of A is (n_samples,n_features), shape of b is (n_samples, n_targets), shape of x is (n_targets, n_features). If $\alpha$ is to be passed as an array, according to the docs, its shape must be (n_targets,). In our case, n_targets is 1. 

I am unable to write alpha array as [0, 0, $\alpha$, $\alpha$] for our test case.

How do I resolve this?

File: pysindy/optimizers/stlsq.py [hunk 2]
```diff
@@ -125,28 +126,55 @@ def __init__(
...
         self.initial_guess = initial_guess
         self.verbose = verbose
+        self.sparse_ind = sparse_ind
 
     def _sparse_coefficients(self, dim, ind, coef, threshold):
         """Perform thresholding of the weight vector(s)"""
         c = np.zeros(dim)
         c[ind] = coef
-        big_ind = np.abs(c) >= threshold
+        if self.sparse_ind is not None:
```
[the remark this answers is not in the document]
Comment (yb6599) [reply pysindy/optimizers/stlsq.py:135]: I've implemented this, should I commit and push right now or should I first do the ridge regression changes and then commit and push it all?

[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/stlsq.py:135]: In an ideal world, each commit should be the smallest change that does a whole, independent thing, e.g. "Change return type of `foo()` to tuple[Bar]", "Clarified docstring of `foo()`", etc.

In practice, that's not always necessary.  However, here I would definitely do the commit and push now - it allows me to review the new code separately from working on the other issue.

File: pysindy/optimizers/stlsq.py (hunk 1 above)
Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/stlsq.py:167]: Oh good catch!  You might have to implement this yourself, motivated by below, and use linear regression

$$
\\|Ax-b\\|^2 + \alpha \\|x\\|^2 = \\left\\|\\left[\begin{matrix}A\\\\ \alpha I\end{matrix}\\right]x - \\left[\begin{matrix}b\\\\ 0\end{matrix}\\right]\\right\\|
$$

Does this make sense?

Comment (yb6599) [reply pysindy/optimizers/stlsq.py:169]: I tried to do it using Linear Regression, however, both of the tests do not pass and the coef values are entirely different than what they should be. I joined A and $\alpha I$ matrices and added zeros to the b matrix, and performed linear regression on them. For the diagonal matrix test, the coefficients for nonsparse indices also changed and are not equal to one, as they should be. 

I tried a different implementation where I created an alpha-array 'A', which is a diagonal matrix $diag(0, 0, \alpha, \alpha)$ for sparse indices [2, 3].  and pass it in the equation for calcultion the minimum weight for ridge regression - 
$$w = (x^Tx + A)^{-1}x^Ty$$
This approach worked and the test with the nondiagonal matrix passed easily. For the one with the diagonal matrix, the test could only pass if I put max_iters = 1.
After each iteration, the regression calculations only take on the columns which have not been thresholded. So when a coefficient gets thresholded and the calculations move onto the next iteration, the column corresponding to that thresholded coefficient is not used in further computations. This means that when a coefficient is thresholded, I will also need to remove the corresponding index value and make changes to other indices in sparse_ind. Let me know if I should proceed with implementing this or if I should try to implement it using the Linear Regression equation you gave above.

Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/stlsq.py]: My comment about $\alpha I$ was just motivation - yes, I was looking for the $diag(0, 0, \alpha, \alpha)$.  Or more specifically, you only need the last two rows.  And - my mistake - these should be $\sqrt{\alpha}$ if directly appended to the feature matrix:

$$
\\left\\|\\left[\begin{matrix}A\\\\  \\left[\begin{matrix}0 & 0 & \sqrt{\alpha} & 0 \\\\ 0 & 0 & 0 & \sqrt{\alpha}\end{matrix} \\right]     \end{matrix}\\right]x - \\left[\begin{matrix}b\\\\ 0 \\\\ 0\end{matrix}\\right]\\right\\|^2
$$

This gives you normal equations of $(A^TA + diag(0, 0, \alpha, \alpha))^{-1}A^Tb$, or in your notation, $(x^T x + A)^{-1}x^Ty$ .  Let me know if that passes the test.

I would prefer that you manage to use scikitlearn for the actual regression.  $(x^T x + A)^{-1}x^Ty$ is correct, but practically, can be solved in ways that do not involve taking the (expensive) inverse.  Using an existing regression toolkit allows someone else to choose how to solve it - QR, pivoted LU, etc.

Comment (yb6599) [reply pysindy/optimizers/stlsq.py]: There are 2 approaches I used for this: normal Linear Regression and Ridge Regression with $\alpha=0$. These two are supposed to be equivalent approaches but I get different coefficient arrays for both of them, the ridge regression with $\alpha=0$ being the one that passes all the tests. I do not understand why there is a difference in the coef array for the supposedly same approaches. I've pushed the updated code and tested it against many different inputs and it works for all of them, it's just with ridge regression with $\alpha=0$ and not normal Linear Regression.

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]

File: pysindy/optimizers/stlsq.py [hunk 3]
```diff
@@ -184,6 +188,15 @@ def _no_change(self):
             last_coef = np.zeros_like(this_coef)
         return all(bool(i) == bool(j) for i, j in zip(this_coef, last_coef))
 
+    def _sparse_remove(self, coef, sparse_sub):
+        """Removes the sparse indices that have been thresholded
+        for subsequent iterations"""
```
[the remark this answers is not in the document]
Comment (yb6599) [reply pysindy/optimizers/stlsq.py]: What would a test for the Remove and Decrement Function look like? I was thinking of checking if the values to remove are given in ascending order and then that the size of existing values gets reduced by the number of values to remove given, is there more I can add here?

[the remark this answers is not in the document]
Comment (Jacob-Stevens-Haas) [reply pysindy/optimizers/stlsq.py]: It could just be a pretty simple test, like the example you gave on paper.

File: pysindy/optimizers/stlsq.py [hunk 4]
```diff
@@ -145,7 +145,7 @@ def _sparse_coefficients(self, dim, ind_nonzero, coef, threshold):
             c[~big_ind] = 0
         return c, big_ind
 
-    def _regress(self, x, y, dim, sparse_sub):
+    def _regress(self, x, y, dim, existing_vals):
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:148]: Another NIT... here, the variable makes sense with the name `sparse_sub`, as this is a function that knows it's role in representing a subset of indices to sparsify.

`_remove_and_decrement`, on the other hand, deals with a generic 1-D array of values.  So within that scope, it calls the argument `existing_vals`

The naming of variables reflects what role the variable plays within the scope, which also helps organize code and tests.

File: pysindy/optimizers/stlsq.py [hunk 5]
```diff
@@ -211,7 +214,7 @@ def _reduce(self, x, y):
         n_samples, n_features = x.shape
         n_targets = y.shape[1]
         n_features_selected = np.sum(ind)
-        sparse_sub = np.array(self.sparse_ind)
+        existing_vals = np.array(self.sparse_ind)
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:217]: Likewise, `_reduce` should keep the name `sparse_sub`

File: pysindy/optimizers/stlsq.py [hunk 6]
```diff
@@ -125,28 +130,54 @@ def __init__(
...
-        with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", category=LinAlgWarning)
-            try:
-                coef = ridge_regression(x, y, self.alpha, **kw)
-            except LinAlgWarning:
-                # increase alpha until warning stops
-                self.alpha = 2 * self.alpha
-        self.iters += 1
-        return coef
+        sparse_ind = self.sparse_ind
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:151]: Rather than creating the local variable `sparse_ind`, can you just use `self.sparse_ind` in the code?

File: pysindy/optimizers/stlsq.py [hunk 7]
```diff
@@ -125,28 +130,54 @@ def __init__(
...
+                warnings.filterwarnings("ignore", category=LinAlgWarning)
+                try:
+                    coef = ridge_regression(x, y, self.alpha, **kw)
+                except LinAlgWarning:
+                    # increase alpha until warning stops
+                    self.alpha = 2 * self.alpha
+            self.iters += 1
+            return coef
+        if sparse_ind is not None:
+            coef = np.zeros(dim)
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:163]: I don't believe this line is necessary.  `try...except` behaves differently with `Warning` objects.  Since `Warnings` do not cause execution to break, the entire `try` block will execute, and then the `except` block will execute.  This means that `coef` will always be assigned a value in the `try...except`, and you can delete `coef = np.zeros(dim)`.

File: pysindy/optimizers/stlsq.py [hunk 8]
```diff
@@ -157,6 +188,18 @@ def _no_change(self):
             last_coef = np.zeros_like(this_coef)
         return all(bool(i) == bool(j) for i, j in zip(this_coef, last_coef))
 
+    def _remove_and_decrement(
+        self, existing_vals: np.ndarray, vals_to_remove: np.ndarray
+    ) -> np.ndarray:
+        """Remove the sparse indices that have been thresholded
+        for subsequent iterations"""
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:195]: Edit the docstring to the context of the function - nothing here about sparsity, thresholding, iterations, or even that `existing_vals` stores indices.

File: pysindy/optimizers/stlsq.py [hunk 9]
```diff
@@ -125,28 +130,54 @@ def __init__(
...
+        if self.sparse_ind is not None:
+            sparse_ind_mask = np.zeros_like(ind_nonzero)
+            sparse_ind_mask[self.sparse_ind] = True
+            c[~big_ind & sparse_ind_mask] = 0
+        if self.sparse_ind is None:
+            c[~big_ind] = 0
         return c, big_ind
 
-    def _regress(self, x, y):
+    def _regress(self, x, y, dim, sparse_sub):
```
Comment (Jacob-Stevens-Haas) [inline pysindy/optimizers/stlsq.py:148]: Add arg descriptions to docstring and add type annotations.

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]
Almost done!  There's two lines of code that can be deleted, and then it's just docstring edits.

**Comment by Jacob-Stevens-Haas** [comment]
Thanks for the commit.  Moving onto the multi-target case!

File: pysindy/optimizers/stlsq.py [hunk 10]
```diff
@@ -206,7 +206,8 @@ def _reduce(self, x, y):
         n_samples, n_features = x.shape
         n_targets = y.shape[1]
         n_features_selected = np.sum(ind)
-        sparse_sub = np.array(self.sparse_ind)
+        sparse_fix = [np.array(self.sparse_ind)] * y.shape[1]
```
Comment (yb6599) [inline pysindy/optimizers/stlsq.py:209]: I had to create 2 similar list of arrays from sparse_ind, one of them remains fixed and the other changes according to the remove_and_decrement function.

**Review by Jacob-Stevens-Haas — CHANGES_REQUESTED** [review]

**Review by Jacob-Stevens-Haas — APPROVED** [review]
LGTM
